### ISSUE TITLE
Terraform errored state not written.

### ISSUE THREAD

**Issue by gtmtech** [hashicorp/terraform#15688]
Terraform 0.9.8

I ran a terraform apply against an aws_db_instance, creating a new resource. The resource was created in ~20m, but terraform appeared not to notice, and kept going on with "Still creating...."

After 2h30, the following error was yielded:

```
Failed to save state: Failed to upload state: NoCredentialProviders: no valid providers in chain. Deprecated.
	For verbose messaging see aws.Config.CredentialsChainVerboseErrors

Failed to persist state to backend.

The error shown above has prevented Terraform from writing the updated state
to the configured backend. To allow for recovery, the state has been written
to the file "errored.tfstate" in the current working directory.

Running "terraform apply" again at this point will create a forked state,
making it harder to recover.

To retry writing this state, use the following command:
    terraform state push errored.tfstate
```

This is a problem in itself, however, I have raised this bug, because scanning the filesystem, there was no errored.tfstate file to push anyway - it didnt create one.

**Comment by jbardin** [comment]
Hi @gtmtech,

I'm not sure what could have happened here. That error you see only comes after the code that writes the `errored.tfstae` itself, and if there was an error writing _that_ file, it would dump the state to stdout as a final effort. 

Are you running Terraform through any sort of automation that may have changed the working directory?

**Comment by gtmtech** [comment]
No. I believe the actual error is due to assuming a role when using an s3 backed state backend - but a role times-out after 60m, requiring a refresh, which currently I believe terraform doesnt do - so when doing long tasks >1h, it can no longer write the state and you get this error. (best guess) - however none of this explains why there is no errored.tfstate file anywhere (I scanned the entire disk, so I know it is nowhere at all)

**Comment by gtmtech** [comment]
More details on the resource that caused this here > https://github.com/terraform-providers/terraform-provider-aws/issues/1307 - but as mentioned above, I believe the source of the cause to be not refreshing an assumed role for s3 state backend

**Comment by apparentlymart** [comment]
Hi @gtmtech,

@jbardin above was talking specifically about the missing `errored.tfstate` file... internally within Terraform this is part of a sequence of increasingly-drastic things Terraform tries to do to avoid losing the state:

* Persist to the remote backend
* If that write fails, write to `errored.tfstate`
* If writing to `errored.tfstate` fails, print the state to `stdout` so it can at least be copy-pasted somewhere

As James was saying, it's odd that you'd get the error message you saw if Terraform failed to write the file, since Terraform _ought_ to have proceeded to the final try of printing to stdout.

The `errored.tfstate` file gets written to the current working directory of the Terraform process, so it may be somewhere other than the root configuration directory if an explicit path is being provided to `terraform apply` to override the default of looking for configuration in the current working directory.

The issue of assume role credentials timing out is something we've seen and are aware of. Terraform _should_ try to refresh assume role tokens if it was Terraform that requested them in the first place, but if the security token is passed in from outside via an environment variable then Terraform doesn't try to refresh it since often the given role-specific credentials don't have access to the `sts` API. If you're using the S3 backend with `role_arn` set and the credentials aren't being refreshed then that's definitely a bug, separate from the `errored.tfstate` not being present. We can discuss that further in terraform/providers/terraform-provider-aws#1307, since the S3 backend uses the authentication code from the AWS provider so the fix would indeed be there anyway.

**Comment by gtmtech** [comment]
In answer to your comments @apparentlymart ,

1. as mentioned, I scanned the entire hard disk for all tfstate files, and none were errored.tfstate - so it wasnt anywhere. Something in this logic isnt working quite as it should. 

2. I am using S3 backend with role_arn set, so in that other bug, it is terraforms responsibility to handle it
